### PR TITLE
Tests scripts enhancements

### LINKED ISSUE #1949: tests scripts enhancements.
* [ ] the test-matrix script needs to support dev build from the bokeh Binstar channel
* [ ] the interactive-tester script needs to accept an specific location for examples (use-case, a conda env containing a dev build which is shipped with the Examples folder, so I want to pass this path: `/media/datos/miniconda3/envs/test34/Examples/bokeh`)

ping @Maggie-M

### PR DESCRIPTION
**Pull request by fpliger**
Issues: fixes #1949 

It also adds the following improvements:

* interactive tester now have the notion of "session" to store progress of the tests. It store every single example file ran and the results. If it crashes before registering the result it will as for the result of the unregistered test.
* windows compatibility (now test_matrix script works also on window)

### CODE REVIEW

**Comment by damianavila** [comment]
this still does not support accepting specific location for examples, are you planning to add this functionality? (I hope so, :wink:)

**Comment by fpliger** [comment]
jeje.. yes, I've mentioned that on the related issue. :+1:

**Comment by damianavila** [comment]
> jeje.. yes, I've mentioned that on the related issue

Yes... I read the PR before the issue... :-P

File: scripts/interactive_tester.py [hunk 1]
```diff
@@ -111,21 +147,40 @@ def main(testing_ground=None):
...
             fileName = "%s/%s" % (testing_ground, fileName)
         try:
-            command = get_cmd(fileName)
-            opener(fileName, command)
 
-            if results.nolog:
-                # Don't display 'next file' message after opening final file in a dir
-                if index != len(TestFiles)-1:
-                    input("\nPress enter to open next file ")
+            if not fileName in last_session:
```
Comment (bryevdv) [inline scripts/interactive_tester.py]: `snake_case` is used everywhere else thoughout, should remain consistent.

File: scripts/interactive_tester.py [hunk 2]
```diff
@@ -111,21 +147,40 @@ def main(testing_ground=None):
...
+                        input("\nPress enter to open next file ")
+                else:
+                    ErrorReport = test_status()
+                    if ErrorReport:
+                        Log.append("\n\n%s: \n %s" % (fileName, ErrorReport))
+
+                    last_session[fileName] = ErrorReport
+                    save_session(last_session)
             else:
-                ErrorReport = test_status()
```
Comment (bryevdv) [inline scripts/interactive_tester.py]: `snake_case`

File: scripts/interactive_tester.py [hunk 3]
```diff
@@ -111,21 +147,40 @@ def main(testing_ground=None):
...
-                if ErrorReport:
-                    Log.append("\n\n%s: \n %s" % (fileName, ErrorReport))
+                prev_res = last_session[fileName]
+                if prev_res == "TESTING...":
+                    print("RESULT OF %s LAST RUN NOT REGISTERED!!" % fileName)
+                    ErrorReport = test_status()
```
Comment (bryevdv) [inline scripts/interactive_tester.py]: same comment

File: scripts/interactive_tester.py [hunk 4]
```diff
@@ -55,12 +57,18 @@ def get_parser():
                         - mpl
                         - pandas
                         - seaborn
+
+                    It also allows to specify a example custom folder path, like:
```
Comment (damianavila) [inline scripts/interactive_tester.py]: You said the same in line 48-49...

**Comment by damianavila** [comment]
trying to run this I get:
```python
(test34)[damian@damian-S400CA][scripts](pr/1971)$ python interactive_tester.py -l /media/datos/miniconda3/envs/test34/Examples/boke
h/                                                                                                                                 
Running examples in custom location: /media/datos/miniconda3/envs/test34/Examples/bokeh/                                           
cleaning previous session file...                                                                                                  
Traceback (most recent call last):                                                                                                 
  File "interactive_tester.py", line 286, in <module>                                                                              
    clean_session()                                                                                                                
  File "interactive_tester.py", line 117, in clean_session                                                                         
    os.remove(SESSION_FILE)                                                                                                        
FileNotFoundError: [Errno 2] No such file or directory: '/media/datos/Desarrollos/bokeh/scripts/INTERACTIVE_TESTER_SESSION.json' 
```

File: scripts/interactive_tester.py (hunk 4 above)
Comment (fpliger) [inline scripts/interactive_tester.py]: Right.. I added a bad description. Lines 48-49 were already there and meant you can only selected from those "pre-built" locations. I intended to say that now you can also specify a custom valid path. 
I'll improve the docstring..

**Comment by fpliger** [comment]
Ah, ok. Will also handle that case when the session file have never been created.

**Comment by damianavila** [comment]
Still not working for me... I mean, the error is fixed but not testing at all...

```python
(test34)[damian@damian-S400CA][scripts](pr/1971)$ ls /media/datos/miniconda3/envs/test34/Examples/bokeh/                           
app  charts  cli  compat  deploy  embed  glyphs  plotting  README.md                                                               
(test34)[damian@damian-S400CA][scripts](pr/1971)$ python interactive_tester.py -l /media/datos/miniconda3/envs/test34/Examples/boke
h/                                                                                                                                 
Running examples in custom location: /media/datos/miniconda3/envs/test34/Examples/bokeh/                                           
cleaning previous session file...                                                                                                  
OK 
```

**Comment by fpliger** [comment]
It doesn't walk into subdirectories, just the main folder you provide. In your case you don't have any file to test on that folder.

**Comment by damianavila** [comment]
ok... we need to improve the docstrings... for something not used to use it (like me), it does not seems obvious from the docstrings you have to add the folder name in the specific path... eventually it would be nice to have it walking through all the examples, but this is something for later...

**Comment by bryevdv** [comment]
Personally I find it more efficient to do one directory for all browsers at at time, than to do all examples for one browser at a time. Otherwise there is just too much context switching between chart types, notebook/non-notebook, server/non-server. I'd rather get one set of examples "over with". If we add a recursive option I think it should be explicit, with a `-r` flag or something.

**Comment by damianavila** [comment]
OK, some other issue:
```
2015-02-25 11:34:58.848 [NotebookApp] Using existing profile dir: '/home/damian/.ipython/profile_default'                          
2015-02-25 11:34:58.874 [NotebookApp] Serving local MathJax from /home/damian/.ipython/profile_default/static/mathjax/MathJax.js at
 /static/mathjax/MathJax.js                                                                                                        
2015-02-25 11:34:58.920 [NotebookApp] The port 8888 is already in use, trying another random port.                                 
2015-02-25 11:34:58.922 [NotebookApp] Serving notebooks from local directory: /media/datos                                         
2015-02-25 11:34:58.922 [NotebookApp] 0 active kernels                                                                             
2015-02-25 11:34:58.922 [NotebookApp] The IPython Notebook is running at: http://localhost:8889/                                   
2015-02-25 11:34:58.922 [NotebookApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).    
2015-02-25 11:34:58.924 [NotebookApp] CRITICAL | /media/datos/bar.ipynb does not exist                                             
Did the plot(s) display correctly? (y/n) ^C(test34)[damian@damian-S400CA][scripts](pr/1971)$                                       
```

If location is added, you need to use the correct path to get the correct ipynb file...
Also, it would be nice to provide options for --port and --profile here: https://github.com/bokeh/bokeh/blob/task/tester_sessions/scripts/interactive_tester.py#L206
Because heavy users of the IPython notebook (like me) deal with custom profile and ports...

**Comment by fpliger** [comment]
@bryevdv the interactive_tester don't have any visibility on the browser it's using. So right now you can either run all the examples with one browser and then reset `BOKEH_BROWSER` to the next browser to test. Or open every example with the default browes, copy the link, open the other browsers, paste the link and test...
I do agree with you that the former flow is more straightforward because it at least reduces context switches..

So, if I get your comment right, your proposal is that we add a new -r argument that does something like:

```
for browser in [os_supported_browsers]:
    set `BOKEH_BROWSER`=browser
    run_all_examples_in_the_specified_folder()
```

Right?

**Comment by bryevdv** [comment]
No, I mean for 0.7.1 I did this to test one directory with all browsers:
```
$ BOKEH_BROWSER="chrome"  python interactive_tester.py -l server
    [output]

$ BOKEH_BROWSER="safari"  python interactive_tester.py -l server
    [output]

$ BOKEH_BROWSER="firefox"  python interactive_tester.py -l serve
    [output]

```
It was incredibly fast to go through directories this way. I'm not opposed to adding more convenience but this worked perfectly well.

**Comment by bryevdv** [comment]
I guess what I am saying about this is that what I like about this tool is that it is really simple, which makes it flexible to use in whatever order or with whatever browsers people are efficient with. I'd personally prefer to keep it simple and flexible and not impose way of doing things.

**Comment by fpliger** [comment]
Ah, ok. So I fully agree... I'm happy with the current status. I see a small improvement in @damianavila proposal in having an argument to walk into sub folder because it's a good fit in the compat examples but now sure it's worth the effort. In general I can live without it :wink: 

Anyway I think it should go into a separate PR

**Comment by damianavila** [comment]
> In general I can live without it :wink:

I agree... btw, did you see my comment on the IPython problems? I guess is the last thing to solve before getting this one merged.

**Comment by damianavila** [comment]
OK... the notebook options is working well now... I think this could be merged.

**Comment by damianavila** [comment]
Restarted TravisCI, I will merge it after succeed.

**Pull request merged**